src/java.base/share/classes/sun/security/ssl/X509TrustManagerImpl.java

```
@@ -248,6 +249,43 @@ private void checkTrusted(X509Certificate[] chain,
         }
     }
 
+    void checkServerTrusted(X509Certificate[] chain, String authType,
+            QuicTLSEngineImpl quicTLSEngine) throws CertificateException {
+        final SSLSession handshakeSession = quicTLSEngine.getHandshakeSession();
+        final List<byte[]> responseList;
+        final AlgorithmConstraints constraints;
+        // determine the AlgorithmConstraints and the OCSP responses
+        // to be applied when validating the certificate chain
+        if (handshakeSession instanceof ExtendedSSLSession extSession) {
+            final String[] localSupportedSignAlgs =
+                    extSession.getLocalSupportedSignatureAlgorithms();
+            constraints = SSLAlgorithmConstraints.forQUIC(
+                    quicTLSEngine, localSupportedSignAlgs, false);
+            // grab any stapled OCSP responses for use in validation
+            responseList = extSession.getStatusResponses();
+        } else {
+            constraints = SSLAlgorithmConstraints.forQUIC(quicTLSEngine, false);
+            responseList = Collections.emptyList();
+        }
+        final SSLParameters sslParameters = quicTLSEngine.getSSLParameters();
+        final Validator v = checkTrustedInit(chain, authType, false);
+        // do the certificate chain validation
+        final X509Certificate[] trustedChain = v.validate(chain, null,
+                responseList, constraints, authType);
+        if (sslParameters != null && handshakeSession != null) {
```

Looks like this `if` block is not needed: `sslParameters != null` condition is always `true`, and we should throw an exception in the beginning of the method if session is null.

---

Moved earlier in 10e3e3508415b901fe0008da4edb5c4f1e8eb85d